### PR TITLE
GridRNN cell uses tuples for output and states

### LINKED ISSUE #4296: Grid RNN Cell does not support dynamic batch sizes
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

http://stackoverflow.com/questions/38442025/tensorflow-grid-lstm-rnn-typeerror
### Environment info

Operating System: 
Ubuntu 16.04 LTS (GNU/Linux 4.4.0-31-generic x86_64)

Installed version of CUDA and cuDNN: 
-rw-r--r-- 1 root root   322936 Sep 19  2015 /usr/lib/x86_64-linux-gnu/libcudadevrt.a
lrwxrwxrwx 1 root root       16 Mär 30 14:25 /usr/lib/x86_64-linux-gnu/libcudart.so -> libcudart.so.7.5
lrwxrwxrwx 1 root root       19 Mär 30 14:25 /usr/lib/x86_64-linux-gnu/libcudart.so.7.5 -> libcudart.so.7.5.18
-rw-r--r-- 1 root root   383336 Sep 19  2015 /usr/lib/x86_64-linux-gnu/libcudart.so.7.5.18
-rw-r--r-- 1 root root   720192 Sep 19  2015 /usr/lib/x86_64-linux-gnu/libcudart_static.a
lrwxrwxrwx 1 root root       12 Apr 14 17:53 /usr/lib/x86_64-linux-gnu/libcuda.so -> libcuda.so.1
lrwxrwxrwx 1 root root       17 Apr 14 17:53 /usr/lib/x86_64-linux-gnu/libcuda.so.1 -> libcuda.so.361.42
-rw-r--r-- 1 root root 16881416 Mär 23 01:42 /usr/lib/x86_64-linux-gnu/libcuda.so.361.42
lrwxrwxrwx 1 root root       13 Aug 31 13:32 /usr/lib/x86_64-linux-gnu/libcudnn.so -> libcudnn.so.4
lrwxrwxrwx 1 root root       17 Aug 31 13:32 /usr/lib/x86_64-linux-gnu/libcudnn.so.4 -> libcudnn.so.4.0.7
-rwxr-xr-x 1 root root 61453024 Aug 31 13:32 /usr/lib/x86_64-linux-gnu/libcudnn.so.4.0.7
-rw-r--r-- 1 root root 62025862 Aug 31 13:32 /usr/lib/x86_64-linux-gnu/libcudnn_static.a

If installed from binary pip package, provide:
1. A link to the pip package you installed:
https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow-0.10.0rc0-cp35-cp35m-linux_x86_64.whl
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`: '0.10.0rc0'
### What other attempted solutions have you tried?

changed file tensorflow/tensorflow/contrib/grid_rnn/python/ops/grid_rnn_cell.py:

```
  # project input
  if inputs is not None and sum(inputs.get_shape().as_list()) > 0 and len(
      conf.inputs) > 0:
```

to:

```
  # project input
  if inputs is not None and inputs.get_shape()[1] > 0 and len(
      conf.inputs) > 0:
```

fixed the issue, but I'm not sure if it is the intended behavior
### Logs or other output that would be helpful

if not changed the following error occurs:

**\* TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'

-> probably because of sum(inputs.get_shape().as_list())

### PR DESCRIPTION
**Pull request by phvu**
In response to this (https://github.com/tensorflow/tensorflow/issues/2560) and recent changes in LSTMCell, this PR added `state_is_tuple=True` and `output_is_tuple=True` into the constructor of GridRNNCell:

```
> cell = tf.contrib.grid_rnn.Grid2LSTMCell(3, use_peepholes=True)
> cell.state_size
       (LSTMStateTuple(c=3, h=3), LSTMStateTuple(c=3, h=3))
> cell.output_size
       (3,)
```

This means there are 2 LSTM cells in the grid, and the state of each cell has size of `(c=3, h=3)`. There is only one output dimension, whose size is 3.

In contrast:

```
> cell = tf.contrib.grid_rnn.Grid2BasicRNNCell(3)
> cell.state_size
     (3, 3)
> cell.output_size
     (3,)
```

This means there are 2 BasicRNN cell in the grid, and the state of each cell has size of 3. There is only one output dimension, whose size is 3.

Previous behaviour is maintained by using `state_is_tuple=False, output_is_tuple=False` when creating the cell.

```
> cell = tf.contrib.grid_rnn.Grid2LSTMCell(3, use_peepholes=True, state_is_tuple=False, output_is_tuple=False)
    WARNING:tensorflow:<tensorflow.contrib.grid_rnn.python.ops.grid_rnn_cell.Grid2LSTMCell object at 0x10de68f50>: Using a concatenated state is slower and will soon be deprecated.  Use state_is_tuple=True.
    WARNING:tensorflow:<tensorflow.contrib.grid_rnn.python.ops.grid_rnn_cell.Grid2LSTMCell** object at 0x10de68f50>: Using a concatenated output is slower and will soon be deprecated.  Use output_is_tuple=True.

> cell.state_size
    12
> cell.output_size
    3

> cell = tf.contrib.grid_rnn.Grid2BasicRNNCell(3, state_is_tuple=False, output_is_tuple=False)
    WARNING:tensorflow:<tensorflow.contrib.grid_rnn.python.ops.grid_rnn_cell.Grid2BasicRNNCell object at 0x10de4b950>: Using a concatenated state is slower and will soon be deprecated.  Use state_is_tuple=True.
    WARNING:tensorflow:<tensorflow.contrib.grid_rnn.python.ops.grid_rnn_cell.Grid2BasicRNNCell object at 0x10de4b950>: Using a concatenated output is slower and will soon be deprecated.  Use output_is_tuple=True.

> cell.state_size
   6
> cell.output_size
   3
```

This also fixes https://github.com/tensorflow/tensorflow/issues/4296

### CODE REVIEW

File: tensorflow/contrib/grid_rnn/python/ops/grid_rnn_cell.py [hunk 1]
```diff
@@ -132,74 +171,118 @@ def __call__(self, inputs, state, scope=None):
...
+          m_prev[recurrent_dim] = array_ops.slice(state, [0, start_idx],
+                                                  [-1, conf.num_units])
+    return c_prev, m_prev, cell_output_size
+
+  def _project_input(self, inputs, c_prev, m_prev, with_c):
+    """Fills in c_prev and m_prev with projected input, for input dimensions
+    """
+    conf = self._config
+
+    if inputs is not None and inputs.get_shape().with_rank(2)[1].value > 0 \
```
Comment (StefOe) [inline tensorflow/contrib/grid_rnn/python/ops/grid_rnn_cell.py]: This indeed fixes #4296

**Comment by ebrevdo** [comment]
Jenkins, test this please.

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by vrv** [comment]
@ebrevdo do you give an LGTM?

**Review by ebrevdo — CHANGES_REQUESTED** [review]

**Comment by ebrevdo** [comment]
there are many changes in this code.  before a more thorough review, can you tell me if this is fully backwards compatible with the original code?  if not, how?

**Comment by phvu** [comment]
Sorry I screwed up this PR (lack of water and sleep apparently). Will close this and submit another PR, with responds to your comments.

**Pull request closed without merging**